android/tangram/jni/jniExports.cpp

```diff
@@ -143,5 +146,88 @@ extern "C" {
         }
     }
 
-}
+    JNIEXPORT jlong JNICALL Java_com_mapzen_tangram_MapController_nativeAddDataSource(JNIEnv* jniEnv, jobject obj, jstring name) {
+        auto source_name = stringFromJString(jniEnv, name);
+        auto source = std::shared_ptr<Tangram::DataSource>(new Tangram::ClientGeoJsonSource(source_name, ""));
+        Tangram::addDataSource(source);
+        return reinterpret_cast<jlong>(source.get());
+    }
+
+    JNIEXPORT void JNICALL Java_com_mapzen_tangram_MapController_nativeRemoveDataSource(JNIEnv* jniEnv, jobject obj, jlong pointer) {
+        assert(pointer > 0);
+        auto source_ptr = reinterpret_cast<Tangram::DataSource*>(pointer);
+        Tangram::removeDataSource(*source_ptr);
+    }
+
+    JNIEXPORT void JNICALL Java_com_mapzen_tangram_MapController_nativeClearDataSource(JNIEnv* jniEnv, jobject obj, jlong pointer) {
+        assert(pointer > 0);
+        auto source_ptr = reinterpret_cast<Tangram::DataSource*>(pointer);
+        Tangram::clearDataSource(*source_ptr, true, true);
+    }
+
+    JNIEXPORT void JNICALL Java_com_mapzen_tangram_MapController_nativeAddFeature(JNIEnv* jniEnv, jobject obj, jlong pointer,
+        jdoubleArray jcoordinates, jintArray jrings, jobjectArray jproperties) {
+
+        assert(pointer > 0);
+        auto source_ptr = reinterpret_cast<Tangram::ClientGeoJsonSource*>(pointer);
 
+        size_t n_points = jniEnv->GetArrayLength(jcoordinates) / 2;
+        size_t n_rings = (jrings == NULL) ? 0 : jniEnv->GetArrayLength(jrings);
+        size_t n_properties = (jproperties == NULL) ? 0 : jniEnv->GetArrayLength(jproperties) / 2;
+
+        Tangram::Properties properties;
+
+        for (size_t i = 0; i < n_properties; ++i) {
+            jstring jkey = (jstring) (jniEnv->GetObjectArrayElement(jproperties, 2 * i));
+            jstring jvalue = (jstring) (jniEnv->GetObjectArrayElement(jproperties, 2 * i + 1));
```

Comment:
But how `GetObjectArrayElement` is freed? It seems that it allocates memory. Referring to discussions about this on stackoverflow and other codebases this is usually what is being done (https://github.com/spurious/SDL-mirror/blob/master/src/main/android/SDL_android_main.c#L45_L52).